File: examples/java/inventory/src/main/java/dev/angzarr/examples/inventory/InventoryState.java
package dev.angzarr.examples.inventory;

import java.util.HashMap;
import java.util.Map;

public record InventoryState(
    String productId,
    int onHand,
    int reserved,
    int lowStockThreshold,
    Map<String, Integer> reservations
) {
    public static InventoryState empty() {
        return new InventoryState("", 0, 0, 0, new HashMap<>());
    }

    public boolean exists() {
        return !productId.isEmpty();
    }

    public int available() {
        return onHand - reserved;
    }
}

File: examples/kotlin/inventory/src/main/kotlin/dev/angzarr/examples/inventory/InventoryState.kt
package dev.angzarr.examples.inventory

data class Reservation(
    val orderId: String,
    val quantity: Int
)

data class InventoryState(
    val productId: String = "",
    val onHand: Int = 0,
    val reserved: Int = 0,
    val reservations: List<Reservation> = emptyList(),
    val lowStockThreshold: Int = 10
) {
    fun exists(): Boolean = productId.isNotEmpty()
    fun available(): Int = onHand - reserved

    companion object {
        fun empty(): InventoryState = InventoryState()
    }
}
